### PR TITLE
GCLOUD: Adopt diff2

### PR DESCRIPTION
**Pull request by tlimoncelli**

### CODE REVIEW

**Comment by tlimoncelli** [comment]
CC @riyadhalnur

File: providers/gcloud/gcloudProvider.go [hunk 1]
```diff
@@ -288,21 +290,27 @@ func (g *gcloudProvider) GetDomainCorrections(dc *models.DomainConfig) ([]*model
 	}
 
 	chg := &gdns.Change{Kind: "dns#change"}
+	const maxGroup = 2000
```
Comment (riyadhalnur) [inline providers/gcloud/gcloudProvider.go:293]: Is this based on a limit set by the gcloud API?

Comment (tlimoncelli) [reply providers/gcloud/gcloudProvider.go:293]: No, I was just guessing numbers.

**Comment by tlimoncelli** [comment]
The tests are failing still.  I think the CREATE isn't working right but I don't know enough about the Google API to correct it.  Can @riyadhalnur take a look?

**Comment by riyadhalnur** [comment]
Will take a lot at it

**Comment by riyadhalnur** [comment]
@tlimoncelli tests are passing. Seems like the formatting step is failing - https://app.circleci.com/pipelines/github/StackExchange/dnscontrol/1728/workflows/9452b352-9cc4-4de3-9393-f113ed18fdcc/jobs/11945/steps?invite=true#step-108-2

**Comment by tlimoncelli** [comment]
The automated tests don't enble `-diff2`.  So, the automated tests are passing but the `-diff2 stuff isn't being tested.

```
❯ go test -v -verbose -provider GCLOUD -diff2 -end 1
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/dnscontroltest-gcloud.com
=== RUN   TestDNSProviders/dnscontroltest-gcloud.com/Clean_Slate:Empty
DEBUG: CHG aBRS l="dnscontroltest-gcloud.com" t="NS" len(m)=2, len(d)=4
DEBUG: gcloud msgs=[DELETE dnscontroltest-gcloud.com NS ns1.example.com. DELETE dnscontroltest-gcloud.com NS ns2.example.com.]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:NS
DEBUG: len-log=105
    integration_test.go:224: DELETE dnscontroltest-gcloud.com NS ns1.example.com.
        DELETE dnscontroltest-gcloud.com NS ns2.example.com.
    integration_test.go:228: runChange error: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'dnscontroltest-gcloud.com. (NS)' already exists, alreadyExists
=== RUN   TestDNSProviders/dnscontroltest-gcloud.com/00:Protocol-Plain:Create_an_A_record
DEBUG: CHG aBRS l="dnscontroltest-gcloud.com" t="NS" len(m)=2, len(d)=4
DEBUG: ADD aBRS l="dnscontroltest-gcloud.com" t="A" len(m)=1, len(d)=1
DEBUG: gcloud msgs=[DELETE dnscontroltest-gcloud.com NS ns1.example.com. DELETE dnscontroltest-gcloud.com NS ns2.example.com.]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:NS
DEBUG: gcloud CREATE
DEBUG: gcloud msgs=[CREATE dnscontroltest-gcloud.com A 1.1.1.1]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:A
DEBUG: len-log=148
    integration_test.go:224: DELETE dnscontroltest-gcloud.com NS ns1.example.com.
        DELETE dnscontroltest-gcloud.com NS ns2.example.com.
        CREATE dnscontroltest-gcloud.com A 1.1.1.1
    integration_test.go:228: runChange error: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'dnscontroltest-gcloud.com. (NS)' already exists, alreadyExists
=== RUN   TestDNSProviders/dnscontroltest-gcloud.com/Post_cleanup:Empty
DEBUG: CHG aBRS l="dnscontroltest-gcloud.com" t="NS" len(m)=2, len(d)=4
DEBUG: gcloud msgs=[DELETE dnscontroltest-gcloud.com NS ns1.example.com. DELETE dnscontroltest-gcloud.com NS ns2.example.com.]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:NS
DEBUG: len-log=105
    integration_test.go:224: DELETE dnscontroltest-gcloud.com NS ns1.example.com.
        DELETE dnscontroltest-gcloud.com NS ns2.example.com.
    integration_test.go:228: runChange error: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'dnscontroltest-gcloud.com. (NS)' already exists, alreadyExists
=== RUN   TestDNSProviders/dnscontroltest-gcloud.com/01:Protocol-Wildcard:Create_wildcard
DEBUG: CHG aBRS l="dnscontroltest-gcloud.com" t="NS" len(m)=2, len(d)=4
DEBUG: ADD aBRS l="*.dnscontroltest-gcloud.com" t="A" len(m)=1, len(d)=1
DEBUG: ADD aBRS l="www.dnscontroltest-gcloud.com" t="A" len(m)=1, len(d)=1
DEBUG: gcloud msgs=[DELETE dnscontroltest-gcloud.com NS ns1.example.com. DELETE dnscontroltest-gcloud.com NS ns2.example.com.]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:NS
DEBUG: gcloud CREATE
DEBUG: gcloud msgs=[CREATE *.dnscontroltest-gcloud.com A 1.2.3.4]
DEBUG: gcloud name:type=*.dnscontroltest-gcloud.com.:A
DEBUG: gcloud CREATE
DEBUG: gcloud msgs=[CREATE www.dnscontroltest-gcloud.com A 1.1.1.1]
DEBUG: gcloud name:type=www.dnscontroltest-gcloud.com.:A
DEBUG: len-log=197
    integration_test.go:224: DELETE dnscontroltest-gcloud.com NS ns1.example.com.
        DELETE dnscontroltest-gcloud.com NS ns2.example.com.
        CREATE *.dnscontroltest-gcloud.com A 1.2.3.4
        CREATE www.dnscontroltest-gcloud.com A 1.1.1.1
    integration_test.go:228: runChange error: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'dnscontroltest-gcloud.com. (NS)' already exists, alreadyExists
=== RUN   TestDNSProviders/dnscontroltest-gcloud.com/Post_cleanup:Empty#01
DEBUG: CHG aBRS l="dnscontroltest-gcloud.com" t="NS" len(m)=2, len(d)=4
DEBUG: gcloud msgs=[DELETE dnscontroltest-gcloud.com NS ns1.example.com. DELETE dnscontroltest-gcloud.com NS ns2.example.com.]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:NS
DEBUG: len-log=105
    integration_test.go:224: DELETE dnscontroltest-gcloud.com NS ns1.example.com.
        DELETE dnscontroltest-gcloud.com NS ns2.example.com.
    integration_test.go:228: runChange error: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'dnscontroltest-gcloud.com. (NS)' already exists, alreadyExists
--- FAIL: TestDNSProviders (2.40s)
    --- FAIL: TestDNSProviders/dnscontroltest-gcloud.com (2.01s)
        --- FAIL: TestDNSProviders/dnscontroltest-gcloud.com/Clean_Slate:Empty (0.45s)
        --- FAIL: TestDNSProviders/dnscontroltest-gcloud.com/00:Protocol-Plain:Create_an_A_record (0.39s)
        --- FAIL: TestDNSProviders/dnscontroltest-gcloud.com/Post_cleanup:Empty (0.36s)
        --- FAIL: TestDNSProviders/dnscontroltest-gcloud.com/01:Protocol-Wildcard:Create_wildcard (0.40s)
        --- FAIL: TestDNSProviders/dnscontroltest-gcloud.com/Post_cleanup:Empty#01 (0.41s)
=== RUN   TestDualProviders
    integration_test.go:333: Clearing everything
DEBUG: CHG aBRS l="dnscontroltest-gcloud.com" t="NS" len(m)=2, len(d)=4
DEBUG: gcloud msgs=[DELETE dnscontroltest-gcloud.com NS ns1.example.com. DELETE dnscontroltest-gcloud.com NS ns2.example.com.]
DEBUG: gcloud name:type=dnscontroltest-gcloud.com.:NS
    integration_test.go:327: #1: DELETE dnscontroltest-gcloud.com NS ns1.example.com.
        DELETE dnscontroltest-gcloud.com NS ns2.example.com.
    integration_test.go:329: runChange error: googleapi: Error 409: The resource 'entity.change.additions[0]' named 'dnscontroltest-gcloud.com. (NS)' already exists, alreadyExists
--- FAIL: TestDualProviders (0.55s)
FAIL
exit status 1
FAIL	github.com/StackExchange/dnscontrol/v3/integrationTest	3.312s
```

**Comment by riyadhalnur** [comment]
@tlimoncelli Thanks for clearing it up. Will take a look into it

**Comment by riyadhalnur** [comment]
@tlimoncelli Sorry for getting back to you with a slight delay (still in a holiday mood). The issue with the failing tests seems to be how we are creating the create requests against the API, for example, to add 2 new servers to an existing NS record, gcp requires you to send the `additions` and `deletions`  arrays in the request.

At the moment, we are sending 
```json
{
	"additions": [
		{
			"kind": "dns#resourceRecordSet",
			"name": "dnscontrol-gcloud.com.",
			"rrdatas": [
				"ns-cloud-c1.googledomains.com.",
				"ns-cloud-c2.googledomains.com.",
				"ns-cloud-c3.googledomains.com.",
				"ns-cloud-c4.googledomains.com.",
				"ns1.example.com.",
				"ns2.example.com."
			],
			"ttl": 300,
			"type": "NS"
		}
	],
	"kind": "dns#change"
}
```

What we should be sending
```json
{
	"additions": [
		{
			"kind": "dns#resourceRecordSet",
			"name": "dnscontrol-gcloud.com.",
			"rrdatas": [
				"ns-cloud-c1.googledomains.com.",
				"ns-cloud-c2.googledomains.com.",
				"ns-cloud-c3.googledomains.com.",
				"ns-cloud-c4.googledomains.com.",
				"ns1.example.com.",
				"ns2.example.com."
			],
			"ttl": 300,
			"type": "NS"
		}
	],
	"deletions": [
		{
			"kind": "dns#resourceRecordSet",
			"name": "dnscontrol-gcloud.com.",
			"rrdatas": [
				"ns-cloud-c1.googledomains.com.",
				"ns-cloud-c2.googledomains.com.",
				"ns-cloud-c3.googledomains.com.",
				"ns-cloud-c4.googledomains.com."
			],
			"ttl": 300,
			"type": "NS"
		}
	],
	"kind": "dns#change"
}
```

**Comment by tlimoncelli** [comment]
(Just seeing this now).  Ah!  Good to know!  I'll take a look at this sometime this week.

**Comment by tlimoncelli** [comment]
I'm going to be lazy... i mean efficient... and use the backwards compatibility mode.  After a few attempts I haven't been able to debug this particular situation.  See https://github.com/StackExchange/dnscontrol/pull/1982 for the new code.

**Pull request closed without merging**